design/backup-performance-improvements.md

```diff
@@ -0,0 +1,211 @@
+- Add `backupItemBlock` return values to `backedUpGroupResources` map
+
+Note that if there are BIAv2 plugins present, we will downgrade all v3 plugins to v2, meaning that we will not have a `GetAdditionalItems` func to call. In this case, we will make the following change to the ItemCollector iteration workflow described above:
+- At the point where we call `GetAdditionalItems` on the item's registered plugins, we will not call this. Instead, we will treat the item as if it had returned no additional items. The ItemBlock will only have one entry.
+
+#### New func `backupItemBlock`
+
+Method signature for new func `backupItemBlock` is as follows:
+```
+func backupItemBlock(block ItemBlock) []schema.GroupResource
+```
+The return value is a slice of GRs for resources which were backed up. Velero tracks these to determine which CRDs need to be included in the backup. Note that we need to make sure we include in this not only those resources that were backed up directly, but also those backed up indirectly via additional items BIA execute returns.
+
+In order to handle backup hooks, this func will first take the input item list (`block.items`) and get a list of included pods, filtered to include only those not yet backed up (using `block.itemBackupper.backupRequest.BackedUpItems`). Iterate over this list and execute pre hooks (pulled out of `itemBackupper.backupItemInternal`) for each item.
+Now iterate over the full list (`block.items`) and call `backupItem` for each. After the first, the later items should already have been backed up, but calling a second time is harmless, since the first thing Velero does is check the `BackedUpItems` map, exiting if item is already backed up). We still need this call in case there's a plugin which returns something in `GetAdditionalItems` but forgets to return it in the `Execute` additional items return value. If we don't do this, we could end up missing items.
+
+After backing up the items in the block, we now execute post hooks using the same filtered item list we used for pre hooks, again taking the logic from `itemBackupper.backupItemInternal`).
+
+#### `itemBackupper.backupItemInternal` cleanup
+
+After implementing backup hooks in `backupItemBlock`, hook processing should be removed from `itemBackupper.backupItemInternal`.
+
+### Phase 2: Process ItemBlocks for a single backup in multiple threads
+
+#### New input field for number of ItemBlock workers
+
+The velero installer and server CLIs will get a new input field `itemBlockWorkerCount`, which will be passed along to the `backupReconciler`.
+The `backupReconciler` struct will also have this new field added. 
+
+#### Worker pool for item block processing
+
+A new type, `ItemBlockWorker` will be added which will manage a pool of worker goroutines which will process item blocks, a shared input channel for passing blocks to workers, and a WaitGroup to shut down cleanly when the reconciler exits.
```

Comment:
if different worker pools are operating somewhat in isolation. 
Could there be scenarios where 1 item is backed up twice somehow?
Just thinking out loud

Reply:
@anshulahuja98 That shouldn't happen here. There are 2 things being done to prevent that:
1) For items in the list returned from the item collector, those are being separated into distinct ItemBlocks, and no item will be in the same block.
2) There is a case where an item is created by a BIA and returned as an additional item - this won't be included in the new GetAdditionalItems API call return value, since it doesn't exist during pre-processing. Normally, that item will *not* be shared among multiple ItemBlocks -- for example, a VS created by the PVC plugin, so this doesn't cause a problem. However, it's possible that there's a shared new item created. For example, say you have CRDs "Foo" and "Bar", and your Foo BIA creates a shared Bar if it doesn't already exist, and returns it as an additional item. Multiple Foos in the backup mean that the shared Bar is returned in more than one place as an AdditionalItem. However, this is solved by Velero's existing code -- there's a map of BackedUpItems that gets added to at the beginning of each BackupItem call -- if the item is already in the map, BackupItem returns immediately. The change needed to support worker pools is that access to this map must be synchronized so that it is thread-safe.

Reply:
got it, thanks for explaining in depth.